### ISSUE TITLE
display.init binds to X11 client instead of console.

### ISSUE THREAD

**Issue by MakerMatrix** [pygame/pygame#2107]
Is there a way to reliably force pygame to render to the console (screen hooked to HDMI, NOT running X11) with e.g. the directfb driver, even when the program is launched from a remote tty?

I was trying to get that to work from a putty session and it seems like it will only render to my X display (I'm tunneled) and if I "unset DISPLAY" then display.init() just fails.

I have to launch the program from the console (pts/0) or one of the virtual console ttys (<ctrl>-<alt>-<fn[2-6]>), in order for it to render there.  Is that as designed or is this user error?  I scoured the docs and examples and didn't find anything quite like this.  There is some language there suggesting that it will choose a display based on permissions, etc, but I guess I'd like to know exactly how it decides which display to use, find out if there are methods to instruct it to override defaults when there are multiple displays it could choose as in my case.  For example is there a display number associated with the console or is that only referring to the number that goes into the X11 DISPLAY variable?

**Comment by MyreMylar** [comment]
I can't quite figure out the setup here.

Are you ssh-ing into a Raspberry Pi that has two displays connected - but one of the displays isn't running your OS? I'm not familiar with that sort of setup. 

As far as I know the `display` parameter of `set_mode()` defaults to 0 which should be the 'primary' display on your system, so usually if you want it to pick the other one in a two display setup you would put `1` here.

Setting the video driver is done with by adding an environment variable before you import pygame like so:

```
import os
os.environ['SDL_VIDEODRIVER'] = 'directfb'
import pygame

pygame.init()

pygame.display.set_mode((800, 600), 0, 0, 1)

```

As an aside pygame 1.9.4 is now a fairly old version of pygame so I can't speak to how accurate my current, imperfect, understanding of pygame is to what it was like back then.

**Comment by MakerMatrix** [comment]
Thanks for the reply.  I will try your suggestions.  Have already recognized that the packaged version of PyGame is very old so have switched to pip3 to manage the python packages.

In this instance, both displays are connected.  For the sake of argument they are a DSI touchscreen and a high-dpi HDMI screen.  The end-goal is to have a "client" app running on the DSI touchscreen (or even remotely on an entirely different computer), which tells the HDMI display what to draw, when, and for how long.  I was originally wanting directfb (no X11) on the HDMI screen but am coming around to other options.  Mostly because there seems to be a relative dearth of knowledge (or at least easily found reference material) around rendering to a console with something like OpenGL ES and surely any "wrappers" around it.

So part of my issue is that I have yet to find a good resource that discusses the various parts of the Raspberry Pi display system(s) and how they interact.  Part of that challenge is that it seems there are various software layers that can be used - layers that I'm not familiar with.   In theory I know how to get X11 to do my bidding.  For the rest of it, I find it a bit confusing, even as a veteran UNIX/Linux admin.

**Comment by robertpfeiffer** [comment]
I got this to work before, but I don't remember everything I had to do.

On the the RPi, you may want to try using os.environ['SDL_VIDEODRIVER'] = 'rpi' instead of directfb and make sure your SDL2 is compiled to link against evdev directly if you want touchscreen input without X11.

pygame 1.9.4 uses SDL1.2 and might not be able to talk to the raspberry pi driver at all.

Your best bet is probably compiling SDL2, SDL_ttf, SDL_mixer, and SDL_image **from source** and enabling the functionality you need. The SDL in the raspbian repos is probably too old and not configured the way you want. Then you definitely need to re-compile pygame 2.0-dev to link against the right SDL2 on your system, not the raspbian SDL2.

See also:
https://hg.libsdl.org/SDL/file/355a4f94a782/docs/README-linux.md
https://hg.libsdl.org/SDL/file/355a4f94a782/docs/README-raspberrypi.md
https://hg.libsdl.org/SDL/file/355a4f94a782/docs/README-directfb.md

**Comment by robertpfeiffer** [comment]
Somebody else got this to work recently, and I got rendering with rpi driver to work by configuring SDL2 with --enable-video-rpi and evdev/udev drivers.

*However* now I look at your question, it looks like you have two screens connected, and we know from that other issue that's something the rpi driver cannot do.

**Comment by MakerMatrix** [comment]
I've come back around to this again.  Same overall issue with the pip3 version - I cannot get a display surface on framebuffer.  It will work with X11.

However I have more experience with pygame this time around so can poke a bit.  I have tried your SDL_VIDEODRIVER='rpi' suggestion and doing that does seem to get me farther.  display.init() returns, as does display.get_driver() (it returns "RPI"), and I'm able to run display.Info() and the returned object indeed shows the correct resolution for the (single) screen I have hooked to this Pi 3B.  So, something sort of works.

However, as soon as  I try to create a Surface with set_mode, it tells me "Could not create  GLES window surface."  If I comment out the SDL_VIDEODRIVER='rpi' line then I can get an X11 surface.

For now I can go back to the packaged version (I think it's 1.9.4) that ships with whatever the latest raspberry pi os and make headway.  That package is actually working this time out (maybe because I've only got one hardware display, or maybe something else in the OS has been fixed in the meantime), but if possible I'd like to get hardware blitting if I can.  My app will do lots of alpha blending on huge (4k-8k) images.

I'll look into compiling SDL2 from source.  What is the aim here?  Uninstall all the packaged SDL2 stuff, build from source, put it in /usr/local and LD_LIBRARY_PATH, and then run with the pip3 version of pygame?

**Comment by MakerMatrix** [comment]
Man, I built all that software, built pygame against it, and have more or less the same problem.  Cannot initiate a display surface.  I get an error about EGL now, though (not GLES).  Saw some articles about using the KMS driver but when I try that the stack just tells me it isn't available (even though I compiled it into SDL2).

At this point I'm pretty sure this is not a PyGame issue, but rather a SDL2 on RPi console display issue.  Really frustrated right now.

**Comment by robertpfeiffer** [comment]
> At this point I'm pretty sure this is not a PyGame issue, but rather a SDL2 on RPi console display issue. Really frustrated right now.

Yes, it definitely is. But I think we should keep this open until we have pi wheels that just work for this type of use case.

**Comment by MakerMatrix** [comment]
I have made significant progress learning about the underpinnings of accelerated FB on Pi.  Will post something, shortly, once I've worked all the kinks out and have pygame running on top of it.

**Comment by robertpfeiffer** [comment]
I think you are now deeper into the issue than all other pygame users or members of the pygame developer team. I only have a RPi3B+, and I know VC4 is sufficiently different, but raspbian devs have assured the public that everything we need for VC4 on raspbian can be downloaded from the official repo.
I'd like to get to the bottom of this, but at this point this seems pi4-specific, so all I can do is ask you to keep us posted, or do a Kickstarter to get Pi4s for the pygame team...

**Comment by MakerMatrix** [comment]
> I'd like to get to the bottom of this, but at this point this seems pi4-specific, so all I can do is ask you to keep us posted, or do a Kickstarter to get Pi4s for the pygame team...

Pi3 should do just fine as long as you use standard display resolutions that are officially supported on that hardware (16:9 1080P or below).  My life is difficult mostly because I need display formats that require a pixel clock higher than the default maximum of 162MHz on Pi3.

It might still be possible to fix that on Pi3, with a custom kernel that overrides max_pixel_clock in vc4_hdmi.c.  But I don't have the time or inclination to test that right now.

I still need to figure out the lib situation now that my display driver issues are understood, and working on Pi4.
